bindings/c/fdb_c.cpp

```diff
@@ -333,6 +340,97 @@ extern "C" DLLEXPORT fdb_error_t fdb_future_get_granule_summary_array(FDBFuture*
 	                 *out_count = na.size(););
 }
 
+void setBlobFilePointer(FDBBGFilePointer* dest, const BlobFilePointerRef& source) {
+	dest->filename_ptr = source.filename.begin();
+	dest->filename_length = source.filename.size();
+	dest->file_offset = source.offset;
+	dest->file_length = source.length;
+	dest->full_file_length = source.fullFileLength;
+}
+
+void setBGMutation(FDBBGMutation* dest, int64_t version, const MutationRef& source) {
+	dest->version = version;
+	dest->type = source.type;
+	dest->param1_ptr = source.param1.begin();
+	dest->param1_length = source.param1.size();
+	dest->param2_ptr = source.param2.begin();
+	dest->param2_length = source.param2.size();
+}
+
+void setBGMutations(FDBBGMutation** mutationsOut, int* mutationCountOut, Arena& ar, const GranuleDeltas& deltas) {
+	// convert mutations from MutationsAndVersionRef to single mutations
+	int mutationCount = 0;
+	for (auto& it : deltas) {
+		mutationCount += it.mutations.size();
+	}
+	*mutationCountOut = mutationCount;
+
+	if (mutationCount > 0) {
+		*mutationsOut = new (ar) FDBBGMutation[mutationCount];
+		mutationCount = 0;
+		for (auto& it : deltas) {
+			for (auto& m : it.mutations) {
+				setBGMutation(&((*mutationsOut)[mutationCount]), it.version, m);
+				mutationCount++;
+			}
+		}
+		ASSERT(mutationCount == *mutationCountOut);
+	}
+}
+
+extern "C" DLLEXPORT fdb_error_t fdb_future_readbg_get_descriptions(FDBFuture* f,
+                                                                    FDBBGFileDescription** out,
+                                                                    int* desc_count) {
+	CATCH_AND_RETURN(Standalone<VectorRef<BlobGranuleChunkRef>> results =
+	                     TSAV(Standalone<VectorRef<BlobGranuleChunkRef>>, f)->get();
+	                 *desc_count = results.size();
```

Comment:
Other fdb_future_get_* functions just return pointers to the data held by the future, and here new memory is allocated on every  fdb_future_readbg_get_descriptions call. Can't we pack the necessary array into the future?

Reply:
It's possible, but it'd be non-trivial to do so. It also gets complicated when we would potentially want to evolve this struct over time in a backwards compatible way

Reply:
The alternative would be to change the implementation of ITransaction::readBlobGranulesStart to return a future with the type as expect by the API (i.e. the stable type). If we want to support evolution, the implementation of DLTransaction::readBlobGranulesStart needs to be changed to use fdb_future_readbg_get_descriptions to extract the value from the returned future and create a new future with it, like it is done in the implementation of other DLTransaction methods, we cannot the future returned by the external client to ThreadSingleAssignmentVar<Standalone<VectorRef<BlobGranuleChunkRef>>>*. I understand that it is an extra effort, but otherwise we are breaking the current design considerations for future compatibility. This can be done in a separate PR if you prefer, but I think we cannot leave it as it is now.

Reply:
Agreed, and let's handle in a separate PR.